### PR TITLE
Encoder options being passed during path based saves

### PR DESCRIPTION
**Pull request by tocsoft**
### Prerequisites

- [x] I have written a descriptive pull-request title
- [x] I have verified that there are no overlapping [pull-requests](https://github.com/JimBobSquarePants/ImageSharp/pulls) open
- [x] I have verified that I am following matches the existing coding patterns and practise as demonstrated in the repository. These follow strict Stylecop rules :cop:.
- [x] I have provided test coverage for my change (where applicable)

### Description

The file path based save operations where failing to pass along the IEncoderOptions. This PR ensure they do and add tests to verify that all the base save operations pass along as expected.

I also Introduced a simple IFileSystem interface to easy testing the file system based codepaths.

### CODE REVIEW

**Comment by tocsoft** [comment]
@dlemstra If I don't have the virtual `SaveInternal()` then I can't verify in my tests that all the same methods are doing the right thing in terms of defaults and passing options along etc. (well not anywhere nearly as easily)... I'll have a think to see if I can find a cleaner/different way to do this without the virtual method.

#### RE: Resetting the stream position.
I just copied that from the first same methods I migrated which was [`public Image<TColor> Save(Stream stream, IImageEncoder encoder, IEncoderOptions options)`](https://github.com/JimBobSquarePants/ImageSharp/pull/135/files#diff-ccb32d856499767c70b37d5e6ad54137L406) where originally we where resetting it back to 0 which I thought was really wrong so instead I reset back to start position. I'll remove this and leave it at the end of the stream.

**Comment by antonfirsov** [comment]
@tocsoft how about an internal `IImageTestCallbacks` interface and an optional `internal Image<T>.TestCallbacks` property?

It's a common issue with TDD, that it has an effect on the design, but I generally support everything, that improves testability! We need to find clever ways to restrict these effects.

**Comment by antonfirsov** [comment]
Looks good to me. @dlemstra do you approve this?

File: src/ImageSharp/IO/IFileSystem.cs [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+
+namespace ImageSharp.IO
+{
+    using System.IO;
+
+ #if !NETSTANDARD1_1
+    /// <summary>
+    /// A simple interface representing the filesystem.
+    /// </summary>
+    public interface IFileSystem
```
Comment (JimBobSquarePants) [inline src/ImageSharp/IO/IFileSystem.cs]: Love the flexibility this brings!

**Comment by dlemstra** [comment]
Not sure if I like this way of testing. We are now testing if private method calls are done correctly. I rather have a test that makes sure that the options are being. We could probably do that by setting IgnoreMetadata to true/false and check the size of the resulting stream?

**Comment by tocsoft** [comment]
@dlemstra I've figured a way to do away to alter my tests to do away with the callbacks class.

**Comment by dlemstra** [comment]
Much better now. You can probably move the code from `SaveInternal` back to `Save`.

**Pull request merged**